h2o-core/src/main/java/water/rapids/RadixOrder.java

```diff
@@ -48,7 +48,8 @@ public void compute2() {
     // it when aligning two keys in Merge()
     int keySize = ArrayUtils.sum(_bytesUsed);
     // 256MB is the DKV limit.  / 2 because we fit o and x together in one OXBatch.
-    int batchSize = 256*1024*1024 / Math.max(keySize, 8) / 2 ;
+  //  int batchSize = 256*1024*1024 / Math.max(keySize, 8) / 2 ; // original value
+    int batchSize = 10000 ; // larger, requires more memory with less remote row fetch and vice versa for smaller
```

Comment:
This is the actual fix ? I don't see any memory consumption changes in the Radix sort.

Reply:
- Isn't this change going to actually slow down the algorithms on machines with plenty of memory ?
- Shouldn't we calculate this dynamically based on available memory (whole cluster has to be taken into account) ?

Reply:
Pavel:

The batchsize of RadixOrder.java will not affect memory consumption in this case.  We are still going to look at the complete sort_cols before determining the orders.  However, you are looking at very few columns and hence, reading the whole frame here is fine.  

The actual change occurs in BinaryMerge.java where we know the correct row order and where the rows are in terms of chunk index and node index.  In this case, I am changing the code so that it will only fill up one batch at a time.  In this case, you are reading in a batch of the frame that contains ALL the columns.  Here, the batchsize matters.  I am setting the batchsize to be just one number.

Small batchsize will generate many batches and hence, you are collecting batches more often from all the nodes/chunks.  This can slow down the sort.  However, in normal data frame, you do not expect to see duplicate values inside a row and hence the total number of batches are mostly one.  It is when you hit the large duplicated rows that you will run into the slow down.  However, if you set the batchsize to be too big, your operation may not complete as the machine will run out of memory for the thread.  My change does not affect the merge operation.

I did a small experiment off the Hadoop and found this using Lauren dataset:
BatchSize, with 10GB Hadoop |   | Original with 30GB Hadoop
-- | -- | --
pow(2,20)=1048576 | pow(2, 14) = 16384 |  
65.13598084 | 77.55316401 | 80.4890089
68.44324303 | 74.94142485 | 84.53140998
65.77826309 | 74.94142485 | 69.37319779
71.68452096 | 77.12972713 | 64.26113796
66.98057103 | 80.23118186 | 64.36953998
68.62468004 | 75.68329501 | 68.41652203
70.24415207 | 76.96755791 | 67.93818879
  | 79.98730803 | 71.41601896
  |   |  
68.12734444 | 77.17938545 | 71.34937805

Note that the slow down is not too sensitive to batchsize as long as it is not too small.  To make things easy, I will choose between 16384 or above.  I will do another set of profiling with merge to just make sure nothing was affected.  Is this clear to you?

Reply:
Pavel:

I thought about using dynamic batchsize but choose not to do it.  The reason is because a RPC is called for merging the portion of the frame with the same MSB.  Since I do not know how many MSBs there are in the frame, I do not know the number of RPCs that will be invoked.  Hence, with different machine cores, memory sizes, this just gets too complicated.  Instead I performed a small measurement experiment and choose the batch size from there.  Once the batch size is determined, I did another set of experiments to measure the performance of sort and merge before and after my change and here are the results:


Merge run time with 50GB |   |  
-- | -- | --
rel-xia | arrange |  
729.1819911 | 463.903742 |  
719.8644481 | 458.936582 |  
702.2259722 | 461.642022 |  
642.1265712 | 460.872243 |  
649.1899869 | 462.709643 |  
631.3483431 | 459.360536 |  
640.1016922 | 458.373506 |  
636.0594559 | 458.688039 |  
657.282526 | 462.229692 |  
766.295264 | 457.157273 |  
677.3676251 | 461.453146 |  
677.3676251 | 460.48422 | Average(s)
sort with 30G |   |  
rel-xia | arrange |  
18.12790608 | 14.62477 |  
18.40748 | 14.68489 |  
21.90696502 | 14.25093 |  
19.80440092 | 14.93063 |  
19.18315411 | 14.84824 |  
18.22094703 | 15.56067 |  
17.98486495 | 16.17139 |  
21.51360393 | 16.32791 |  
20.96344399 | 19.08546 |  
19.32334495 | 13.82125 |  
19.5436111 | 15.4306141 | Average(s)

Reply:
Since the two results are comparable in orders of magnitude, I consider my implementation did not slow down the sort/merge by too much.  The hadoop cluster changes all the time and I consider the results comparable if they have the same order of magnitude.  Does this help alleviate your concern?  Let me know.  Thanks, Wendy

Reply:
@wendycwong Of course, the slow down I was talking about is within the same order of magnitude and thus we can potentially ignore it. This is good enough. Definitely an improvement over what we had previously.

Having the value dynamically set would be benificial - there will be no slowdown (if not even performance increase) for machines with large memory. I understand your reasons. It would be too complicated to do right now and there is no real benefit.